internal/trie/node/header.go

```diff
@@ -11,24 +11,35 @@ import (
 
 // encodeHeader writes the encoded header for the node.
 func encodeHeader(node *Node, writer io.Writer) (err error) {
+	if node == nil {
+		_, err = writer.Write([]byte{emptyVariant.bits})
+		return err
+	}
+
 	partialKeyLength := len(node.PartialKey)
 	if partialKeyLength > int(maxPartialKeyLength) {
 		panic(fmt.Sprintf("partial key length is too big: %d", partialKeyLength))
 	}
 
 	// Merge variant byte and partial key length together
-	var variant variant
+	var nodeVariant variant
 	if node.Kind() == Leaf {
-		variant = leafVariant
+		if node.HashedValue {
+			nodeVariant = leafWithHashedValueVariant
+		} else {
+			nodeVariant = leafVariant
+		}
 	} else if node.StorageValue == nil {
-		variant = branchVariant
+		nodeVariant = branchVariant
+	} else if node.HashedValue {
+		nodeVariant = branchWithHashedValueVariant
 	} else {
-		variant = branchWithValueVariant
+		nodeVariant = branchWithValueVariant
 	}
```

Comment:
can we use a `switch` here?

Reply:
Let me try but we are checking different things here